### ISSUE TITLE
Spire server doesn't work after Postgres switch leader

### ISSUE THREAD

**Issue by kaige0207** [spiffe/spire#4565]
* **Version**:  v1.7.2 
* **Platform**: redhat8 
* **Subsystem**: Server plugin: DataStore "sql"

I deployed a spire server cluster and its **DataStore** plugin uses a Postgres cluster. My Postgres cluster has three nodes, one leader node and two replica nodes. Only the leader node can create connections that are both readable and writable, but the spire server does not work after the leader node is switched. I found that the old connections were not released and the new connections were not established.

Log:
> level=error msg="Failed to reload configs" error="datastore-sql: pq: cannot set transaction read-write mode during recovery" subsystem_name=bundle_client

I want to add the `target_session_attrs=read-write` parameter to the Postgres connection string but the spire server does not support it.

Log:
> level=error msg="Fatal run error" error="datastore-sql: pq: unrecognized configuration parameter \"target_session_attrs\""

 
I checked the Posrgres related code in spire server and found that the **Gorm** library we used was very low version. I guess the database driver does not support it.

**Comment by evan2645** [comment]
@kaige0207 are you configuring multiple hostnames into the connection string, or are you using a single hostname that resolves to multiple addresses? If you can provide an anonymized example of your connection string, that would be awesome. Thank you!

**Comment by kaige0207** [comment]
@evan2645 Yes, I using a single hostname that resolves to multiple addresses. I using Envoy to forward the connection and I solved this problem. Since the time for Postgres cluster to switch leader nodes is often less than the time for envoy health check, this will cause envoy to connect to the replica node and make Spire unavailable. I solved this problem by closing the connections on the replica node using _pg_terminate_backend_ function. 
Of course, if the connection string of Spire supports the `target_session_attrs=read-write` parameter, no connections will be created on the replica node at any time. The [pgx ](https://github.com/jackc/pgx) library supports this parameter but the [lib/pq](https://github.com/lib/pq) library does not support it. I found [gorm2 ](https://github.com/go-gorm/gorm) use the pgx driver so if we upgrade gorm this problem will be solved.


> connection_string = "dbname=test user=test host=postgres-test-dns port=8888 password=xxxxx target_session_attrs=read-write sslmode=disable"

Thank you very much for your reply.
